cluster-autoscaler/cloudprovider/aws/aws_cloud_provider.go

```diff
@@ -128,8 +128,18 @@ func (aws *awsCloudProvider) NodeGroupForNode(node *apiv1.Node) (cloudprovider.N
 }
 
 // HasInstance returns whether a given node has a corresponding instance in this cloud provider
-func (aws *awsCloudProvider) HasInstance(*apiv1.Node) (bool, error) {
-	return true, cloudprovider.ErrNotImplemented
+func (aws *awsCloudProvider) HasInstance(node *apiv1.Node) (bool, error) {
+	awsRef, err := AwsRefFromProviderId(node.Spec.ProviderID)
+	if err != nil {
+		return false, err
+	}
+
+	// we don't care about the instance health
+	if _, ok := aws.awsManager.asgCache.instanceStatus[*awsRef]; !ok {
```

Comment:
@vadasambar please use the [`asgCache.InstanceStatus` function](https://github.com/kubernetes/autoscaler/blob/e29dff136fabc9ecf8b3efb03cd7fb847b7686d4/cluster-autoscaler/cloudprovider/aws/auto_scaling_groups.go#L231-L240) instead of directly accessing (in a non-thread-safe fashion) the underlying map :)

Reply:
Thank you for noticing it! I have updated the code.